force-app/main/default/lwc/lookup/lookup.js

```diff
@@ -0,0 +1,422 @@
+            if (this._cleanSearchTerm.length >= this.minSearchTermLength) {
+                // Display spinner until results are returned
+                this.loading = true;
+
+                const searchEvent = new CustomEvent('search', {
+                    detail: {
+                        searchTerm: this._cleanSearchTerm,
+                        rawSearchTerm: newSearchTerm,
+                        selectedIds: this._curSelection.map((element) => element.id)
+                    }
+                });
+                this.dispatchEvent(searchEvent);
+            }
+            this._searchThrottlingTimeout = null;
+        }, SEARCH_DELAY);
+    }
+
+    isSelectionAllowed() {
+        if (this.isMultiEntry) {
+            return true;
+        }
+        return !this.hasSelection();
+    }
+
+    hasSelection() {
+        return this._curSelection.length > 0;
+    }
+
+    processSelectionUpdate(isUserInteraction) {
+        // Reset search
+        this._cleanSearchTerm = '';
+        this._searchTerm = '';
+        this.setSearchResults([...this._defaultSearchResults]);
+        // Indicate that component was interacted with
+        this._isDirty = isUserInteraction;
+        // If selection was changed by user, notify parent components
+        if (isUserInteraction) {
+            const selectedIds = this._curSelection.map((sel) => sel.id);
+            this.dispatchEvent(new CustomEvent('selectionchange', { detail: selectedIds }));
+        }
+    }
+
+    // EVENT HANDLING
+
+    handleInput(event) {
+        // Prevent action if selection is not allowed
+        if (!this.isSelectionAllowed()) {
+            return;
+        }
+        this.updateSearchTerm(event.target.value);
+    }
+
+    handleKeyDown(event) {
+        if (this._focusedResultIndex === null) {
+            this._focusedResultIndex = -1;
+        }
+        if (event.keyCode === KEY_ARROW_DOWN) {
+            // If we hit 'down', select the next item, or cycle over.
+            this._focusedResultIndex++;
+            if (this._focusedResultIndex >= this._searchResults.length) {
+                this._focusedResultIndex = 0;
+            }
+            event.preventDefault();
+        } else if (event.keyCode === KEY_ARROW_UP) {
+            // If we hit 'up', select the previous item, or cycle over.
+            this._focusedResultIndex--;
+            if (this._focusedResultIndex < 0) {
+                this._focusedResultIndex = this._searchResults.length - 1;
+            }
+            event.preventDefault();
+        } else if (event.keyCode === KEY_ENTER && this._hasFocus && this._focusedResultIndex >= 0) {
+            // If the user presses enter, and the box is open, and we have used arrows,
+            // treat this just like a click on the listbox item
+            const selectedId = this._searchResults[this._focusedResultIndex].id;
+            this.template.querySelector(`[data-recordid="${selectedId}"]`).click();
```

Comment:
Since I can't see the definition of `this.template` from doing a `CTRL+F` of `this.template`, I'm wondering if there's ever a need for a guard clause for `!this.template`? Like:

```javascript
if (!this.template) {
  // Short-circuit the function logic
  return;
}
```

Reply:
Just got this from Thomas, for posterity:

> just a FYI for this.template - it refers to the outermost <template> tag in the LWC html file (kind of similar to window object, without it wouldn't even deploy ) so I think we're good without the if(!this.template) check.